diff --git a/tools/legion_prof_rs/src/backend/data_source.rs b/tools/legion_prof_rs/src/backend/data_source.rs
index 74c2a87861..5622599655 100644
--- a/tools/legion_prof_rs/src/backend/data_source.rs
+++ b/tools/legion_prof_rs/src/backend/data_source.rs
@@ -255,16 +255,14 @@ impl StateDataSource {
                 kind_index += 1;
 
                 let color = match (kind, device) {
-                    (ProcKind::GPU, Some(DeviceKind::Device)) => Color::OLIVEDRAB,
-                    (ProcKind::GPU, Some(DeviceKind::Host)) => Color::ORANGERED,
-                    (ProcKind::CPU, None) => Color::STEELBLUE,
-                    (ProcKind::Utility, None) => Color::CRIMSON,
-                    (ProcKind::IO, None) => Color::ORANGERED,
-                    (ProcKind::ProcGroup, None) => Color::ORANGERED,
-                    (ProcKind::ProcSet, None) => Color::ORANGERED,
-                    (ProcKind::OpenMP, None) => Color::ORANGERED,
-                    (ProcKind::Python, None) => Color::OLIVEDRAB,
-                    _ => unreachable!(),
+                    // All Device timelines get the same color
+                    (_, Some(DeviceKind::Device)) => Color::OLIVEDRAB,
+                    // Some processors get special colors
+                    (ProcKind::CPU, _) => Color::STEELBLUE,
+                    (ProcKind::Utility, _) => Color::CRIMSON,
+                    (ProcKind::Python, _) => Color::DARKGOLDENROD,
+                    // Everything else gets the standard color
+                    _ => Color::ORANGERED,
                 };
                 let color: Color32 = color.into();
 
@@ -1006,7 +1004,7 @@ impl StateDataSource {
                                 } else if event.is_barrier() {
                                     item_meta.fields.push(ItemField(
                                                 self.fields.critical,
-                                                Field::String(format!("Waiting on unknown critical path barrier {:#x} created on node {}. Please load the logfile from at least one node that arrives on this barrier to start determining a critical path. You'll need to load the logs from all nodes that arrive on this barrier to determine a precise critical path. If you see this message and did not run with the -lg:prof_all_critical_arrivals flag then please report this case as it is likely a bug.", event.0, event.node_id().0)), 
+                                                Field::String(format!("Waiting on unknown critical path barrier {:#x} created on node {}. Please load the logfile from at least one node that arrives on this barrier to start determining a critical path. You'll need to load the logs from all nodes that arrive on this barrier to determine a precise critical path. If you see this message and did not run with the -lg:prof_all_critical_arrivals flag then please report this case as it is likely a bug.", event.0, event.node_id().0)),
                                                 Some(Color32::BLUE)));
                                 } else {
                                     item_meta.fields.push(ItemField(
