diff --git a/cmake/HighFiveConfig.cmake b/cmake/HighFiveConfig.cmake
index 33ce7b7..a270d4c 100644
--- a/cmake/HighFiveConfig.cmake
+++ b/cmake/HighFiveConfig.cmake
@@ -14,6 +14,9 @@ if(HDF5_IS_PARALLEL)
   find_dependency(MPI)
   target_link_libraries(HighFive::HighFive INTERFACE MPI::MPI_C MPI::MPI_CXX)
 endif()
-
-add_library(HighFive ALIAS HighFive::HighFive)
-add_library(HighFiveInclude ALIAS HighFive::Include)
+if(NOT TARGET HighFive)
+  add_library(HighFive ALIAS HighFive::HighFive)
+endif()
+if(NOT TARGET HighFiveInclude)
+  add_library(HighFiveInclude ALIAS HighFive::Include)
+endif()
