diff --git a/runtime/legion/legion_redop.cc b/runtime/legion/legion_redop.cc
index 9d4cf49da..75ef9c3b3 100644
--- a/runtime/legion/legion_redop.cc
+++ b/runtime/legion/legion_redop.cc
@@ -20,10 +20,10 @@
 namespace Legion {

 #ifdef LEGION_REDOP_HALF
-  /*static*/ const __half SumReduction<__half>::identity = __half(0, false/*raw*/);
-  /*static*/ const __half DiffReduction<__half>::identity = __half(0, false/*raw*/);
-  /*static*/ const __half ProdReduction<__half>::identity = __half(1, false/*raw*/);
-  /*static*/ const __half DivReduction<__half>::identity = __half(1, false/*raw*/);
+  /*static*/ const __half SumReduction<__half>::identity = __half(0);
+  /*static*/ const __half DiffReduction<__half>::identity = __half(0);
+  /*static*/ const __half ProdReduction<__half>::identity = __half(1);
+  /*static*/ const __half DivReduction<__half>::identity = __half(1);
   /*static*/ const __half MaxReduction<__half>::identity = __half(-2e10);
   /*static*/ const __half MinReduction<__half>::identity = __half(2e10);
 #endif
@@ -45,10 +45,10 @@ namespace Legion {

 #ifdef LEGION_REDOP_COMPLEX
 #ifdef LEGION_REDOP_HALF
-  /*static*/ const complex<__half> SumReduction<complex<__half> >::identity = complex<__half>(__half(0, false/*raw*/), __half(0, false/*raw*/));
-  /*static*/ const complex<__half> DiffReduction<complex<__half> >::identity = complex<__half>(__half(0, false/*raw*/), __half(0, false/*raw*/));
-  /*static*/ const complex<__half> ProdReduction<complex<__half> >::identity = complex<__half>(__half(1, false/*raw*/), __half(0, false/*raw*/));
-  /*static*/ const complex<__half> DivReduction<complex<__half> >::identity = complex<__half>(__half(1, false/*raw*/), __half(0, false/*raw*/));
+  /*static*/ const complex<__half> SumReduction<complex<__half> >::identity = complex<__half>(__half(0), __half(0));
+  /*static*/ const complex<__half> DiffReduction<complex<__half> >::identity = complex<__half>(__half(0), __half(0));
+  /*static*/ const complex<__half> ProdReduction<complex<__half> >::identity = complex<__half>(__half(1), __half(0));
+  /*static*/ const complex<__half> DivReduction<complex<__half> >::identity = complex<__half>(__half(1), __half(0));
 #endif
   /*static*/ const complex<float> SumReduction<complex<float> >::identity = complex<float>(0.f, 0.f);
   /*static*/ const complex<float> DiffReduction<complex<float> >::identity = complex<float>(0.f, 0.f);
