diff --git a/runtime/CMakeLists.txt b/runtime/CMakeLists.txt
index 5bd02e9e3..56fbcc181 100644
--- a/runtime/CMakeLists.txt
+++ b/runtime/CMakeLists.txt
@@ -549,6 +549,7 @@ if(Legion_USE_CUDA)
   add_library(realm_cuhook SHARED
     realm/cuda/cuda_hook.cc)
   set_target_properties(realm_cuhook PROPERTIES POSITION_INDEPENDENT_CODE ON)
+  set_target_properties(realm_cuhook PROPERTIES OUTPUT_NAME "realm_cuhook${INSTALL_SUFFIX}")
   if(REALM_LIMIT_SYMBOL_VISIBILITY)
     set_target_properties(realm_cuhook PROPERTIES CXX_VISIBILITY_PRESET hidden)
     set_target_properties(realm_cuhook PROPERTIES VISIBILITY_INLINES_HIDDEN 1)
