diff --git a/realm/src/CMakeLists.txt b/realm/src/CMakeLists.txt
index 5c9d404b0..8eb8c773e 100644
--- a/realm/src/CMakeLists.txt
+++ b/realm/src/CMakeLists.txt
@@ -76,15 +76,15 @@ if(TARGET CUDA::cuda_driver AND REALM_USE_CUDA)
   list(APPEND REALM_CUDA_SOURCES cuda/cuda_memcpy.cu)
 endif()
 
-if(TARGET hip::host)
+if(TARGET hip::host AND REALM_ENABLE_HIP)
   list(APPEND REALM_SOURCES hip/hip_module.cc hip/hip_internal.cc hip/hip_access.cc)
 endif()
 
-if(TARGET LLVM::LLVM)
+if(TARGET LLVM::LLVM AND REALM_ENABLE_LLVM)
   list(APPEND REALM_SOURCES llvmjit/llvmjit_internal.cc llvmjit/llvmjit_module.cc)
 endif()
 
-if(TARGET hdf5::hdf5)
+if(TARGET hdf5::hdf5 AND REALM_ENABLE_HDF5)
   list(APPEND REALM_SOURCES hdf5/hdf5_module.cc hdf5/hdf5_internal.cc hdf5/hdf5_access.cc)
 endif()
 
@@ -99,11 +99,11 @@ if(REALM_USE_PREALM)
   list(APPEND REALM_SOURCES prealm/prealm.cc)
 endif()
 
-if(TARGET Python3::Python)
+if(TARGET Python3::Python AND REALM_ENABLE_PYTHON)
   list(APPEND REALM_SOURCES python/python_module.cc python/python_source.cc)
 endif()
 
-if(TARGET ucx::ucp)
+if(TARGET ucx::ucp AND REALM_ENABLE_UCX)
   list(
     APPEND
     REALM_SOURCES
@@ -118,16 +118,16 @@ if(TARGET ucx::ucp)
   )
 endif()
 
-if(TARGET GASNet::GASNet)
+if(TARGET GASNet::GASNet AND REALM_ENABLE_GASNETEX)
   list(APPEND REALM_SOURCES gasnet1/gasnet1_module.cc gasnet1/gasnetmsg.cc)
   list(APPEND REALM_SOURCES gasnetex/gasnetex_module.cc gasnetex/gasnetex_internal.cc)
 endif()
 
-if(TARGET MPI::MPI_CXX)
+if(TARGET MPI::MPI_CXX AND REALM_ENABLE_MPI)
   list(APPEND REALM_SOURCES mpi/mpi_module.cc mpi/am_mpi.cc)
 endif()
 
-if(TARGET Kokkos::kokkos)
+if(TARGET Kokkos::kokkos AND REALM_ENABLE_KOKKOS)
   list(APPEND REALM_SOURCES kokkos_interop.cc)
   set_source_files_properties("${REALM_SOURCE_DIR}/kokkos_interop.cc" 
       DIRECTORY ${PROJECT_SOURCE_DIR}
