diff --git a/include/experimental/__p0009_bits/extents.hpp b/include/experimental/__p0009_bits/extents.hpp
index 98a57e3..4f783e0 100644
--- a/include/experimental/__p0009_bits/extents.hpp
+++ b/include/experimental/__p0009_bits/extents.hpp
@@ -18,7 +18,7 @@
 #include "dynamic_extent.hpp"
 #include "utility.hpp"
 
-#ifdef __cpp_lib_span
+#if defined(__cpp_lib_span) && (__cpp_lib_span >= 202002L)
 #include <span>
 #endif
 #include <array>
