diff --git a/include/experimental/__p0009_bits/layout_stride.hpp b/include/experimental/__p0009_bits/layout_stride.hpp
index 2e5a745..554824b 100644
--- a/include/experimental/__p0009_bits/layout_stride.hpp
+++ b/include/experimental/__p0009_bits/layout_stride.hpp
@@ -29,7 +29,7 @@
 #include <type_traits>
 #include <utility>
 
-#ifdef __cpp_lib_span
+#if defined(__cpp_lib_span) && (__cpp_lib_span >= 202002L)
 #include <span>
 #endif
 #if defined(_MDSPAN_USE_CONCEPTS) && MDSPAN_HAS_CXX_20 && defined(__cpp_lib_concepts)
